### PR TITLE
Enhanced Hazelcast configuration strategy

### PR DESCRIPTION
**Pull request by dima767**
This PR introduces a smart configuration facility for `HazelcastInstance` used by `HazelcastTicketRegistry` which auto-configures it programmatically by an exposed set of configuration properties (as before), but switches to a more advanced native hazelcast configuration source if such a need arises by reacting to the new `hz.config.location` property set by deployer. More info is available in the updated documentation also included in this PR.

I would really like to have this facility also available in `4.2`, hence all the `since` tags in the code read `4.2.0`.

### CODE REVIEW

**Comment by mmoayyed** [comment]
:+1: looks great. 

Two questions:
- Could we add the new property to cas.properties also, to be consistent? 
- Technically, per our policy RC is in feature-freeze mode and I'd consider this a new feature. Shall we wait and push this patch into 4.2.1? Any chance this could be considered a "bug fix" so that we can justify pushing the change into 4.2?

**Comment by mmoayyed** [comment]
By the way I suspect this patch closes #1161  right?

**Comment by dima767** [comment]
I sure will add that property - that was an overlook on my part. Wrt "bug fix" - sure :-) as there is currently no way to have a varying config pieces, like join strategies, etc. without requiring users to do some twisted bean definition gymnastics with aliases, etc. Actually ran into this during real word client work requiring aws join strategy instead of a default tcp/ip. (Ping @scalding)

**Comment by dima767** [comment]
Oh, that's right, closes #1161

**Comment by mmoayyed** [comment]
Very good. I'd give this a day so that others can review. Otherwise, we'll merge this back into 4.2.

File: cas-server-integration-hazelcast/src/main/java/org/jasig/cas/ticket/registry/config/HazelcastInstanceConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+     * Create HazelcastInstance bean.
+     *
+     * @param hazelcastConfigLocation String representation of hazelcast xml config.
+     * @param resourceLoader resource loader for loading hazelcast xml configuration resource.
+     * @return HazelcastInstance bean.
+     *
+     * @throws IOException if parsing of hazelcast xml configuration fails
+     */
+    @Bean
+    public HazelcastInstance hazelcast(@Value("${hz.config.location:NO_CONFIG_PROVIDED}") final String hazelcastConfigLocation,
```
Comment (leleuj) [inline cas-server-integration-hazelcast/src/main/java/org/jasig/cas/ticket/registry/config/HazelcastInstanceConfiguration.java]: What's the use of this default value: `NO_CONFIG_PROVIDED`?

Comment (leleuj) [inline cas-server-integration-hazelcast/src/main/java/org/jasig/cas/ticket/registry/config/HazelcastInstanceConfiguration.java]: oK. I think I get it: the `ResourceLoader` returns `false` to `exists()` in that case, right?

Comment (dima767) [inline cas-server-integration-hazelcast/src/main/java/org/jasig/cas/ticket/registry/config/HazelcastInstanceConfiguration.java]: Exactly right!

**Comment by leleuj** [comment]
+1

**Comment by dima767** [comment]
Great. I'll add one more commit (add the new props to the cas.properties) as noted by @mmoayyed and we are good to go.

**Comment by mmoayyed** [comment]
Looks like we're done here. Thanks @dima767

**Pull request merged**